core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -719,6 +718,7 @@ protected void doRun() throws Exception {
         private void executeLocally() throws Exception {
             // execute locally
             Tuple<Response, ReplicaRequest> primaryResponse = shardOperationOnPrimary(state.metaData(), request);
+            primaryResponse.v2().primaryTerm(indexShardReference.opPrimaryTerm());
```

Comment:
My goal here is not to force a subclass to worry about these things. As you noted later on - the primary term of a specific primary shard can not change.